modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+         * it generates
+         * \f[q = [\cos\theta, u_x\sin\theta,u_y\sin\theta,  u_z\sin\theta].\f]
+         */
+        Quat(const _Tp angle, const cv::Vec<_Tp, 3> &axis);
+
+        /**
+         * @brief from a 3x3 rotate matrix
+         */
+        explicit Quat(const cv::Mat &R);
+
+        /**
+         * @brief from Rodrigues vector
+         */
+        explicit Quat(const Vec<_Tp, 3> &rvec);
+
+        /**
+         * @brief a way to get element
+         * @param index over a range [0, 3].
+         *
+         * Assume a quaternion q,
+         *
+         * q.at(0) is equicalent to q.w,
+         *
+         * q.at(1) is equicalent to q.x,
+         *
+         * q.at(2) is equicalent to q.y,
+         *
+         * q.at(3) is equicalent to q.z,
+         */
+        _Tp at(size_t index) const;
+
+        /**
+         * @brief return the conjugate of this quaternion
+         * \f[q.conjugate() = (w, -x, -y, -z)\f]
+         */
+        Quat<_Tp> conjugate() const;
+
+        /**
+         * @brief return the value of exponential value
+         * \f[\exp(q) = e^w (\cos||\boldsymbol{v}||+ \frac{v}{||\boldsymbol{v}||})\sin||\boldsymbol{v}||\f]
+         * @param q a quaternion
+         *
+         * For example:
+         * ```
+         * Quatd q{1,2,3,4};
+         * cout << exp(q) << endl;
+         * ```
+         */
+        template <typename T>
+        friend Quat<T> exp(const Quat<T> &q);
+
+        /**
+         * @brief return the value of exponential value
+         * \f[\exp(q) = e^w (\cos||\boldsymbol{v}||+ \frac{v}{||\boldsymbol{v}||}\sin||\boldsymbol{v}||)\f]
+         *
+         * For example
+         * ```
+         * Quatd q{1,2,3,4};
+         * cout << q.exp() << endl;
+         * ```
+         */
+        Quat<_Tp> exp() const;
+
+        /**
+         * @brief return the value of logarithm function
+         * \f[\ln(q) = \ln||q|| + \frac{\boldsymbol{v}}{||\boldsymbol{v}||}\arccos\frac{a}{||q||}\f]
+         * @param q a quaternion
+         * @param assumeUnit if true, q should be a unit quaternion and this function will save some computations.
+         *
+         * For example
+         * ```
+         * Quatd q1{1,2,3,4};
+         * cout << log(q1) << endl;
+         * ```
+         */
+        template <typename T>
+        friend Quat<T> log(const Quat<T> &q, bool assumeUnit);
+
+        /**
+         * @brief return the value of logarithm function
+         *  \f[\ln(q) = \ln||q|| + \frac{\boldsymbol{v}}{||\boldsymbol{v}||}\arccos\frac{w}{||q||}\f]
+         * @param assumeUnit if true, this quaternion should be a unit quaternion and this function will save some computations.
+         *
+         * For example
+         * ```
+         * Quatd q(1,2,3,4);
+         * q.log();
+         *
+         * Quatd q1(1,2,3,4);
+         * q1.normalize().log(true);
+         * ```
+         */
+        Quat<_Tp> log(bool assumeUnit=false) const;
```

Comment:
There are too many methods with `assumeUnit` parameter.

1. It make sense to make it as a template parameter.
2. Or what is about creation of derived class for such quaternions handling?

```
base Quat=> QuatUnit
or
base QuatGeneric => Quat
```

(need to check terminology in articles and other libraries)

Reply:
1. Our original intention of designing this parameter is to speed up some calculations. If a user knows about this param, they'll get an extra potential performance, otherwise the code will just work. 
If we make it as a templeta param, user will have to determine whether it is a unit one when construct a quaternion and other operations. I think this may not be friendly.


2. Before implementing quaternion, Vadim and Rostislav agreed that there is only one class to represent quaternions and unit quaternions.
here is the main content between Rostislav and Vadim
![1017](https://user-images.githubusercontent.com/49099366/96288719-7731a800-1016-11eb-80ca-8441fdb2799f.png)
![101702](https://user-images.githubusercontent.com/49099366/96288742-7f89e300-1016-11eb-885f-f3577da02434.png)

Rostislav to me 
![101703](https://user-images.githubusercontent.com/49099366/96288765-87498780-1016-11eb-8715-c530b44d1533.png)

![101704](https://user-images.githubusercontent.com/49099366/96289226-3c7c3f80-1017-11eb-9f37-08b0355edd3b.png)

Reply:
/cc @vpisarev There are several options:
1. `derived` vs `separate` class
2. template parameter vs just regular parameter.
3. left "as is"

P.S. Mainly, strict typing helps to avoid shooting legs by misused parameter value.

Reply:
It is possible to make a template parameter to give a strategy about unification of the quaternion coefficients with default value as `false`.
```cpp
template <class T, bool IsUnit = false>
class Quat {};
```
If you add a converting constructor that performs conversion between unit and general storage usage of the class will be pretty simple, but unfortunately you have to make `w, x, y, z` private to keep class invariant.

Reply:
Maybe adding a template parameter `assumeUnit` is a good idea. A note should be added for a user that when this flag is on, this would make the code a bit faster but wouldn't preserve the unit norm, it's a user's responsibility.
Also, this would leave class members public.

Reply:
OK

Reply:
During the process of converting assumeUnit to template parameter, I have encountered some problem and I think make  this param to template is not flexible enough.

Here are some reasons:

1. Some functions will change the quaternion from unit to not unit one or from  not unit to unit one. like exp(), log().

   If a user set `Quat<double, true> q{1/2,1/2,1/2,1/2};` and it will become not unit one after `log(q)`， while the flag is still on(true).

   

2. In interpolation function, all inputs must be same including the `assumeUnit` in each quaternion, otherwise it can't match the function. But this is not required in this code.

So I think left assumeUnit in each function is better.

Reply:
@chargerKong there was a meeting between me, @alalek and @vpisarev
We settled on a previous solution (a lot of functions with default `assumeUnit` argument) but with the following additions:
* the `assumeUnit` argument should be enum of two values (like ASSUME_UNIT and ASSUME_NON_UNIT for example), not bool - so that it'd be easier to search it in the code or in the docs
* the set of functions with `assumeUnit` should be revised: sometimes this assumption will not give significant performance improvement (for example, I'm not sure if normalization takes a lot in 'slerp', 'squad' or 'spline' functions).

Reply:
Thanks all for the suggestions!

I totally agree the first suggestions.

The second one can also be done, 
but here are my thoughs to surpport it may be better to keep `assumeUnit` argument in these functions.
 - Adding this param could ensure the uniformity of the code style. So I think users may also expect these functions to have the assumeUnit parameter. 
- It does not improve much performance in some functions, but it actually saves some time, even though the improvement rate is low.

Reply:
OK, for uniformity purposes, let's keep some functions with `assumeUnit` argument.

Reply:
OK. Thanks